### PR TITLE
🟢 Lamarckian bench harness (buildDataset/meanError/trainCreature/perturb) triplicated (Issue #3638)

### LINKED ISSUE #3638: 🟢 Lamarckian bench harness (buildDataset/meanError/trainCreature/perturb) triplicated
The Lamarckian-evolution benchmark harness — a synthetic logistic-teacher dataset generator, the mean-absolute-error scorer, the one-generation backprop trainer (identical `createBackPropagationConfig` block: `plankConstant: 1e-7`, unit adjustment scales, `disableRandomSamples: true`, `batchSize: 1`), and the weight/bias perturbation operator — is copy-pasted across three benchmarks. Severity: low.

Sites:

- `bench/EvolutionPaceLeverComparison.ts:251-327` (`buildDataset` 251-274, `meanError` 276-287, `trainCreature` 289-313, `perturb` 315-327)
- `bench/ProductionPaceLeverBakeOff.ts:222-297` (`buildDataset` 222-245, `meanError` 247-258, `trainCreature` 260-284, `perturb` 286-297)
- `bench/TrainPerGenConvergence.ts:103-167` (`buildDataset` 103-124, `meanError` 126-133, `trainCreature` 135-154, `perturb` 156-167 — a reworded copy: module constants and a bare `rng()` in place of the options object, same rule throughout)

## The shared knowledge

The rule is: *what problem these related pace/convergence experiments are solved against and how a candidate is trained and scored* — teacher targets are `logistic(Σ wᵢxᵢ)` with weights and inputs drawn uniformly from [-1, 1), fitness is mean absolute error, one Lamarckian step is `activateAndTrace`/`propagate` over the dataset for N inner iterations under the exact backprop config above, and perturbation jitters every weight and bias by `±scale`. If any of that changed — a different teacher non-linearity, a corrected backprop setting — all three benchmarks would need the same edit, or their results silently stop being comparable across the experiment family.

## Suggested fix

No existing helper covers this, but with three call sites a small extracted module qualifies, and `bench/` already has precedent for shared harness modules (`bench/score_per_hour_harness.ts`, `bench/Constants.ts`). Extract `bench/lamarckian_harness.ts` exporting roughly:

```ts
export function buildTeacherDataset(rng: () => number, opts: { inputs: number; outputs: number; datasetSize: number }): Sample[];
export function meanAbsoluteError(creature: Creature, data: readonly Sample[], outputs: number): number;
export function trainOneGeneration(creature: Creature, data: readonly Sample[], learningRate: number, innerIters: number): void;
export function perturb(source: Creature, rng: () => number, scale: number): Creature;
```

Each benchmark keeps its own constants and passes them in — a call per site, no per-caller flags. If a future experiment deliberately needs a different teacher, it writes its own generator rather than growing this one.

🏷️ Filed by idle-task template: `duplicated-knowledge` · Run id: `vibe-msbmtovn-f5732c`

### PR DESCRIPTION
**Pull request by stservice**
# Lamarckian bench harness extracted to a shared module (Issue #3638)

## Summary

The Lamarckian-evolution benchmark harness — the synthetic logistic-teacher
dataset generator, the mean-absolute-error scorer, the one-generation backprop
trainer (`plankConstant: 1e-7`, unit adjustment scales, `disableRandomSamples`,
`batchSize: 1`) and the weight/bias perturbation operator — was copy-pasted
across three benchmarks. A corrected backprop setting or a different teacher
non-linearity would have needed the same edit in all three, or their results
would silently stop being comparable across the experiment family.

That shared knowledge now lives once in `bench/lamarckian_harness.ts`, and the
three benches call it, each keeping its own constants and passing them in.
Closes #3638.

- New: `bench/lamarckian_harness.ts` — `buildTeacherDataset`,
  `meanAbsoluteError`, `trainOneGeneration`, `perturb`, and the `Sample` type.
- Updated call sites: `bench/EvolutionPaceLeverComparison.ts`,
  `bench/ProductionPaceLeverBakeOff.ts`, `bench/TrainPerGenConvergence.ts` (~200
  duplicated lines removed).
- Docs: the `buildDataset` reference in `docs/PERFORMANCE_RESEARCH.md` (the
  human-run GRQ adoption-gate recipe) now names `buildTeacherDataset`.

The two benches that drive the RNG through `RandomNumberGenerator` pass
`() => rng.random()`, so the random streams — and hence every reported figure —
are unchanged. `buildTeacherDataset` also validates its problem sizes and throws
a `RangeError` rather than quietly producing an empty dataset (fail loud).

```mermaid
flowchart LR
    H["bench/lamarckian_harness.ts<br/>buildTeacherDataset · meanAbsoluteError<br/>trainOneGeneration · perturb"]
    A[EvolutionPaceLeverComparison.ts] --> H
    B[ProductionPaceLeverBakeOff.ts] --> H
    C[TrainPerGenConvergence.ts] --> H
```

## Evidence

Backend/CLI change — no web interface to screenshot.

**Behaviour is unchanged.** Each benchmark was run before and after the refactor
and the outputs diffed. Every non-timing figure is byte-identical; only
wall-clock milliseconds (inherently variable) differ.

| benchmark                         | before vs after                                  |
| --------------------------------- | ------------------------------------------------ |
| `TrainPerGenConvergence.ts`       | identical (best error, improvement %)            |
| `EvolutionPaceLeverComparison.ts` | identical except the wall-clock ms column        |
| `ProductionPaceLeverBakeOff.ts`   | fully identical (whole stdout, all three sweeps) |

Sample — `TrainPerGenConvergence.ts`, identical before and after:

```
trainPerGen= 1 -> best error 0.050095
trainPerGen= 4 -> best error 0.044458
trainPerGen=10 -> best error 0.034280
```

Quality gate: `./quality.sh < /dev/null` →
`ok | 8146 passed (5 steps) | 0 failed | 4 ignored`.

## Test Plan

New: `test/bench/LamarckianHarness.ts` — nine tests calling the real harness
functions and asserting on outcomes:

- `buildTeacherDataset` — sample/target shape with inputs in `[-1, 1)` and
  logistic targets strictly inside `(0, 1)`; determinism for a seeded RNG;
  `RangeError` on non-positive or non-integer problem sizes (error path).
- `meanAbsoluteError` — zero when the targets are the creature's own outputs;
  equals the mean absolute deviation over samples × outputs.
- `trainOneGeneration` — reduces error on a learnable dataset; leaves the
  creature untouched with zero inner iterations (edge case).
- `perturb` — jitters every weight and bias within `±scale` and leaves the
  source creature unmutated; a zero scale yields an unchanged copy.

Existing `test/bench/EvolutionPaceLeverComparison.ts` and
`test/bench/ProductionPaceLeverBakeOff.ts` (determinism and score-carry
contract) pass unchanged, confirming the extraction preserved behaviour.

---

🤖 Processed by: stservice
🏷️ Run id: `vibe-mscj1s7i-09790e`

### CODE REVIEW

**Review by nleck — APPROVED** [review]

**Comment by stservice** [comment]
I've pushed a fix for the CI failure (**Merge coverage & results**). Please review the changes.

**Comment by stservice** [comment]
I've pushed fixes for the spelling issues. Please review the changes.

**Comment by stservice** [comment]
## CI fix: `Merge coverage & results`

**Root cause — a real test failure, not an infrastructure fault.**

The merge job's own log only says `❌ 1 shard(s) reported test failures`. The shard jobs deliberately never fail on test failures (they record a status marker so every shard still uploads artefacts), so all 8 shards showed green while the gate failed. Tracing the gate output identified **shard 3**, and its `junit-3.xml` artefact carried the two real failures:

```
test/ci/WorkflowVersionCheckStrictMode.ts
  ✗ .github/workflows/publish.yml "Determine whether the version needs publishing"
      run block opens with set -Eeuo pipefail (Issue #3003)
  ✗ .github/workflows/github-release.yml "Read version from deno.json"
      run block opens with set -Eeuo pipefail (Issue #3003)
```

Commit `3d3ae5ed` ("Normalise workflow run-block shell hygiene", Issue #3716) rewrote every workflow `run:` block to `set -euo pipefail` — including the two version-handling blocks that `test/ci/WorkflowVersionCheckStrictMode.ts` pins to `set -Eeuo pipefail` under Issue #3003. Both blocks carry an inline comment forbidding exactly that rewrite:

```yaml
# `-E` is pinned here by test/ci/WorkflowVersionCheckStrictMode.ts
# (Issue #3003) — do not normalise to `set -euo pipefail`.
```

**Fix** — restored `-E` in the two pinned blocks:

- `.github/workflows/publish.yml` — step `needs_publish`
- `.github/workflows/github-release.yml` — step `version`

I checked for a competing gate before reverting: no test in `test/ci/` enforces the blanket `set -euo pipefail` form, so these two blocks are the intended exception to #3716's normalisation and restoring `-E` does not regress it. The other blocks #3716 touched are unchanged.

**Verification** — `./quality.sh` passed locally: **8148 passed, 0 failed, 4 ignored (5m21s)**, including both previously-failing tests.

**Comment by stservice** [comment]
## CI fix — `Merge coverage & results`

**Root cause.** The merge gate failed with `❌ 1 shard(s) reported test failures`; the log only reports the count, so I pulled the shard statuses from the merge job (`shard 3: failed`) and the `shard-3` artefact's `junit-3.xml` for the actual assertions. Two tests in `test/ci/WorkflowVersionCheckStrictMode.ts` failed:

- `.github/workflows/publish.yml "Determine whether the version needs publishing"`
- `.github/workflows/github-release.yml "Read version from deno.json"`

Commit `3d3ae5ed` ("Normalise workflow run-block shell hygiene", Issue #3716) rewrote `set -Eeuo pipefail` → `set -euo pipefail` across all workflow run blocks, including these two — which are explicitly pinned by that test (Issue #3003) and carry an inline `do not normalise to set -euo pipefail` comment directly above the line that was changed.

**Fix.** Restored `set -Eeuo pipefail` in the two pinned blocks:

- `.github/workflows/publish.yml:57`
- `.github/workflows/github-release.yml:35`

The remaining 22 normalised blocks and the 5 newly-added `set` lines from #3716 are untouched — no test pins them and dropping the inert `-E` there is intentional.

**Verification.** `deno test -A test/ci/WorkflowVersionCheckStrictMode.ts` → 2 passed, 0 failed (previously 2 failed). `./quality.sh` passed locally.

Note: the shard job itself is green by design (shards never fail on test failures so every shard uploads its artefact) — the merge gate is the enforcing check, so this was a genuine test failure, not a flake.
